[STAGES 5-6] SAP BTP Multitenancy

polinatonko opened the pull request:
Complete below steps:
1. Create and configure Approuter
2. Implement API that returns tokenUrl, clientId, clientSecret of bound XSUAA service instance (endpoint /api/v1/application-info) and allow access for ADMIN role only
3. Update Approuter’s welcomFile path to endpoint from Step 2
4. Create xs-security.json file with roles templates for XSUAA 
5. Implement API to receive subscription/unsubscription requests from SaaS Provisioning Service that returns tenant-specific approuter URL (https://{{tenant-subdomain}}-{{approuter-name}})
6. Create saas-provisioning.json file with application details for SaaS Provisioning Service to register application in marketplace
7. Implement request filter to set current tenant info (tenantId, subdomain) from JWT into request context
8. Implement logic for new database schema creation/deletion on tenant subscription/unsubscription locally
9. Implement logic for new HANA schema creation/deletion via Service Manager API on tenant subscription/unsubscription in cloud
10. Adapt Liquibase schema migration to multitenant mode
11. Configure connection to required schema using Hibernate
12. Implement data source with connection pools per tenant
13. Adjust logic with Destination service integration to multitenant mode (i.e. try to find destination by name on subscriber subaccount level firstly, if not presented – on provider subaccount level)
14. Implement endpoint /dependencies that returns Destination service xsappname during subscription
15. Create Destination with SMTP server credentials on the Subscriber subaccount level with the same name as at Provider subaccount level

polinatonko committed:
feat: create and configure approuter

polinatonko committed:
feat: add endpoints for application info and subscription callbacks, saas-provisioning.json; update xs-security.json and mta.yaml for multitenancy

polinatonko committed:
feat: add request filter to extract subdomain value from jwt; remove xsuaa authentication from actuator endpoints

polinatonko committed:
feat: implement logic for new database schema creation/deletion on tenant subscription/unsubscription locally

polinatonko committed:
feat: implement logic for new HANA schema creation/deletion via Service Manager API on tenant subscription/unsubscription in cloud

polinatonko committed:
fix: update handling authentication errors for basic authentication

polinatonko committed:
fix: update cache for access tokens

polinatonko committed:
fix: identifying tenant locally and in cloud

polinatonko committed:
feat: adapt liquibase schema migration to multitenant mode

polinatonko committed:
refactor: refactor btp classes

polinatonko committed:
feat: configure connection to required schema using hibernate locally

polinatonko committed:
fix: remove route for actuator endpoints from approuter configuration

polinatonko committed:
chore: add database_id parameter when creating service instance

polinatonko committed:
chore: clean up mta.yaml

polinatonko committed:
fix: tenant subscription response

polinatonko committed:
refactor: refactor api paths

polinatonko committed:
doc: add javadoc for saasprovisioningservice package

polinatonko committed:
refactor: refactor jwt converter

polinatonko committed:
refactor: add hibernate properties customizer class

polinatonko committed:
feat: add saving subdomain into tenant context in cloud

polinatonko committed:
feat: implement multitenancy on the database level

SN4NTR commented on btp/saas-provisioning/config.json:
```diff
@@ -0,0 +1,17 @@
+  ],
+  "displayName": "Learning Management System",
+  "description": "Learning Management System platform",
+  "category": "Test",
```
I suggest just to put these values higher, it's like a title for our document :)
```
{
  "xsappname": "lms",
  "displayName": "Learning Management System",
  "description": "Learning Management System platform",
  "category": "Test",
  "appUrls": {
    "getDependencies": "https://lms-dev.cfapps.us10-001.hana.ondemand.com/api/v1/callback/dependencies",
    "onSubscription": "https://lms-dev.cfapps.us10-001.hana.ondemand.com/api/v1/callback/tenants/{tenantId}"
  },
  "appPlans": [
    {
      "name": "application",
      "description": "Plan for development purposes"
    }
  ],
  "onSubscriptionAsync": false
```

SN4NTR commented on mta.yaml:
```diff
@@ -12,22 +12,57 @@ modules:
       build-result: target/*.jar
+
+    requires:
+      - name: approuter-api
```
approuter-api is not needed for our backend, backend should know nothing about approuter

SN4NTR commented on mta.yaml:
```diff
@@ -12,22 +12,57 @@ modules:
       JBP_CONFIG_JAVA_OPTS: "[java_opts: '-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=8000']"
       JBP_CONFIG_SAP_MACHINE_JDK: "{ version: 21.+ }"
       TARGET_RUNTIME: tomcat
+      APPROUTER_URL: ~{approuter-api/url}
```
this property is also not required

SN4NTR commented on mta.yaml:
```diff
@@ -12,22 +12,57 @@ modules:
+    provides:
+      - name: approuter-api
+        properties:
+          url: ${routes/0/route}
```
approuter-api is not needed

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/config/XSUAAProperties.java:
```diff
@@ -0,0 +1,28 @@
+@ConfigurationProperties(prefix = "sap.security.services.xsuaa")
+@Getter
+@Setter
+public class XSUAAProperties {
```
based on java conventions, naming should be the following: `XsuaaProperties`

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/controller/ApplicationController.java:
```diff
@@ -0,0 +1,20 @@
+@RestController
+@RequestMapping("/api/v1")
+@RequiredArgsConstructor
+public class ApplicationController {
```
it's better name it `ApplicationInfoController`, and `@RequestMapping("/api/v1/application-info")`

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/controller/ApplicationController.java:
```diff
@@ -0,0 +1,20 @@
+    private final XSUAAProperties xsuaaProperties;
+
+    @GetMapping("/application-info")
+    public XSUAAProperties getXSUAAProperties() {
```
it's better to create and return `ApplicationInfoDto` here. ApplicationInfoDto should contain `XsuaaProperties` inside
Then method should be renamed to `getInfo()`

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @GetMapping("/dependencies")
+    public List<ServiceInfoDto> getDependencies(@RequestParam("tenantId") String tenantId) {
```
`tenantId` is not provided for this request as I know

polinatonko replied:
It doesn't provide much value, but as we can see from the application logs, tenantId is specified: "GET /api/v1/callback/dependencies?tenantId=1ebd6692-faab-417d-b472-614f574f53e0 HTTP/1.1"

Info can be found here: https://community.sap.com/t5/technology-blog-posts-by-members/understanding-dependencies-in-saas-provisioning/ba-p/13484556

SN4NTR replied:
then it's fine :)

SN4NTR commented on src/main/java/org/example/learningsystem/core/security/converter/JwtConverter.java:
```diff
@@ -0,0 +1,53 @@
+            log.info("Jwt Roles: {}", roles);
+            authorities.addAll(roles);
+        }
+        return new JwtAuthenticationToken(source, authorities);
```
you can rewrite it with Optional.ofNullable().map()...

SN4NTR commented on src/main/resources/application-cloud.yaml:
```diff
@@ -5,12 +5,18 @@ spring:
     password: ${vcap.services.lms-hana-schema.credentials.password}
 
+  main:
+    allow-bean-definition-overriding: true
```
why it's needed in cloud only?

SN4NTR commented on src/main/resources/application-cloud.yaml:
```diff
@@ -30,4 +36,18 @@ mail:
+sap:
+  security:
+    services:
+      xsuaa:
```
then it's better to place all BTP services envs here too

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/config/ServiceDependencies.java:
```diff
@@ -0,0 +1,23 @@
+@ConfigurationProperties(prefix = "application")
+@Getter
+@Setter
+public class ServiceDependencies {
```
it's better to rename class where `Application` will be a part of class name or rename environment variable to contain `service`, because class should represent what exactly we are mapping from envs

SN4NTR commented on approuter/xs-app.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "authenticationMethod": "route",
```
approuter should redirect us our application info endpoint after login:
```
"welcomeFile": "/api/v1/application-info"
```

polinatonko committed:
refactor: refactor stage-5

polinatonko committed:
feat: adjust logic with Destination service integration to multitenant mode

polinatonko committed:
refactor: refactor multitenancy

polinatonko committed:
chore: update postman collection

polinatonko committed:
docs: add javadoc to classes in the servicemanager package

polinatonko committed:
docs: add javadoc to classes from the multitenancy package

polinatonko committed:
chore: remove unused rest client method

polinatonko committed:
docs: add javadoc to the xsuaa btp package

polinatonko committed:
Merge branch 'final-stage-5' into stage-6

polinatonko committed:
fix: edit service manager uri builder

polinatonko committed:
docs: update postman collection

polinatonko committed:
refactor: usage of optional class and lombok annotations

polinatonko committed:
refactor: data source configuration

polinatonko committed:
refactor: database schema management locally

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/v1/callback")
```
and maybe it would be even better to change path to `/api/v1/subscriptions` :)
```
PUT /api/v1/subscriptions/tenants/{id}
DELETE /api/v1/subscriptions/tenants/{id}
GET  /api/v1/subscriptions/dependencies
```

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,48 @@
+package org.example.learningsystem.btp.saasprovisioningservice.controller;
```
I suggest to place this controller in package `org.example.learningsystem.multitenancy.subscription.controller` (and everything related to multitenancy here)
As it's not explicitly related to SaaS Provisioning service, this API can be triggered by anyone who has `Callback` scope, SubscriptionService can be implemented in different ways based on environment where tenant should be subscribed

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/controller/ApplicationInfoController.java:
```diff
@@ -0,0 +1,21 @@
+package org.example.learningsystem.btp.saasprovisioningservice.controller;
```
this controller is related to another feature and can be placed in `org.example.learningsystem.application.controller` package. Purpose of this controller to return any kind of data about application (not even related to BTP) so it should be abstract as much as possible

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/controller/ApplicationInfoController.java:
```diff
@@ -0,0 +1,21 @@
+@RequiredArgsConstructor
+public class ApplicationInfoController {
+
+    private final XsuaaProperties xsuaaProperties;
```
In order to be abstract, it's better to create `ApplicationInfoService` that will return `ApplicationInfoDto` and based on implementation we can either return XsuaaProperties inside or something else even not related to BTP

SN4NTR commented on src/main/java/org/example/learningsystem/btp/saasprovisioningservice/service/SubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,40 @@
+@RequiredArgsConstructor
+public class SubscriptionServiceImpl implements SubscriptionService {
+
+    private static final String TENANT_SPECIFIC_URL_TEMPLATE = "https://%s-lms-approuter-dev.cfapps.us10-001.hana.ondemand.com";
```
approuter name shouldn't be hardcoded and should be moved to environment variables and it can be placed in user-provided service (as it can be changed)
Additionally, space and domain can be changed and we should be prepared for this. It also can be done via environment variables and we can read it from VCAP_APPLICATION env (you can find in service bindings of your app) `${vcap.application.application_uris[0]}`
And then you can build tenant-specific URL:
```
String approuterUri = applicationUri.replace(applicationName, approuterName);
return "%s://%s-%s".formatted(HTTPS_PROTOCOL, tenantSubdomain, approuterUri);
```

SN4NTR commented on src/main/java/org/example/learningsystem/core/security/converter/JwtConverter.java:
```diff
@@ -0,0 +1,59 @@
+package org.example.learningsystem.core.security.converter;
```
I recommend to place this converter in `org.example.learningsystem.btp.xsuaa.converter` as it's XSUAA-specific converter (other auth servers can generate JWT with another structure and we will have to implement dedicated converter for it)

SN4NTR commented on src/main/resources/application-cloud.yaml:
```diff
@@ -30,4 +22,25 @@ mail:
\ No newline at end of file
+  start-tls-enable: ${vcap.services.lms-user-service.credentials.startTls}
+
+sap:
```
it's better to call env like:
```
btp:
  services:
    destination-service:
      uri: ...
    feature-flags:
      uri: ...
    ...
```

polinatonko committed:
refactor: stage-5

polinatonko committed:
refactor: refactor stage-6

polinatonko committed:
fix: several small fixes

Update the request context filter configuration. Add closure of
tenants' data sources on application shutdown. Add exception handling
to the email service. Minor refactoring.

polinatonko committed:
chore: remove odd user-provided-service

polinatonko committed:
refactor: minor refactoring

polinatonko committed:
Merge branch 'stage-6' into final-stage-5

polinatonko committed:
fix: local setup

polinatonko committed:
docs: update postman collection

polinatonko committed:
style: edit formatting

polinatonko committed:
fix: closure of tenant data sources

polinatonko committed:
fix: delete custom converter

polinatonko committed:
chore: update security configuration

polinatonko committed:
fix: security configuration, open api docs

polinatonko committed:
chore: refactor multitenant data source

polinatonko committed:
Merge branch 'stage-5' of https://github.com/polinatonko/LearningSystem into stage-5

polinatonko committed:
[STAGE-6.1] Additional Tasks

Add CHANGELOG.md file with description of changes based on project version
Update README.md with project description, stack and local setup
Add Application Autoscaler for application (cpuutil metric)
Configure sending emails with course reminders asynchronously (Async annotation and custom thread pool)
Update roles and scopes configuration
Place basic authentication configuration in user-provided service in cloud

SN4NTR commented on src/main/java/org/example/learningsystem/btp/destinationservice/service/CloudDestinationServiceImpl.java:
```diff
@@ -29,37 +35,56 @@ public class CloudDestinationServiceImpl implements DestinationService {
+            } catch (NotFound e) {
+                log.info("Destination {} wasn't found for current tenant", name);
+            }
+        }
```
if you're checking `Optional.isPresent()` or similar, it means that it could be rewritten somehow (otherwise sense of  this wrapper is lost):
```
var tenantTokenUrlOpt = getTenantTokenUrl();
return tenantTokenUrlOpt.map(token -> ...);
```

Ideally, code should be rewritten to:
```
    public Destination getByName(String name) {
        return findDestinationOnSubscriberLevel(name)
                .or(() -> findDestinationOnProviderLevel(name))
                .orElseThrow(() -> new LearningManagementSystemException("Destination is not found [name = %s].".formatted(name)));
    }

    private Optional<Destination> findDestinationOnSubscriberLevel(String name) {
        Supplier<OAuth2ClientCredentials> clientCredentialsSupplier = this::buildClientCredentialsForSubscriber;
        return tryToGetDestination(name, clientCredentialsSupplier);
    }

    private Optional<Destination> findDestinationOnProviderLevel(String name) {
        Supplier<OAuth2ClientCredentials> clientCredentialsSupplier = this::buildClientCredentialsForProvider;
        return tryToGetDestination(name, clientCredentialsSupplier);
    }

    private Optional<Destination> tryToGetDestination(String name, Supplier<OAuth2ClientCredentials> clientCredentialsSupplier) {
        try {
            OAuth2ClientCredentials clientCredentials = clientCredentialsSupplier.get();
            return Optional.ofNullable(destinationServiceClient.getByName(name, clientCredentials));
        } catch (Exception e) {
            log.error("Failed to get destination [name = %s].".formatted(name), e);
            return Optional.empty();
        }
    }

```

SN4NTR commented on src/main/java/org/example/learningsystem/btp/servicemanager/instance/service/ServiceInstanceManager.java:
```diff
@@ -0,0 +1,67 @@
+    private final BaseServiceManager baseServiceManager;
+    private final String databaseId;
+    private final ServiceManagerRestClient serviceManagerRestClient;
+    private final ServiceManagerUriBuilder serviceManagerURIBuilder;
```
serviceManagerURIBuilder -> serviceManagerUriBuilder

SN4NTR commented on src/main/java/org/example/learningsystem/multitenancy/db/datasource/MultiTenantDataSource.java:
```diff
@@ -0,0 +1,98 @@
+    private final DataSource defaultDataSource;
+    private final TenantDataSourceManager tenantDataSourceManager;
+
+    private Map<Object, Object> targetDataSources;
```
it's better to define key as `TenantInfo` and value as `DataSource` (because you are using only these types in your map)

polinatonko committed:
refactor: stages 5-6

Extract the logic of interacting with the Destination Service API to the
DestinationServiceClient. Refactor MultitenantDataSource. Add a base
exception class.

polinatonko committed:
refactor: rename variables

polinatonko committed:
refactor: classes throughout the project

polinatonko committed:
refactor: rename variables, refactor xsuaa url provider

Move Xsuaa URI template to the User-provided service.
Rename variables, edit log messages.

polinatonko committed:
docs: add javadoc to multitenancy package

polinatonko committed:
refactor: change multitenancy package structure

polinatonko committed:
refactor: package names, error messages

polinatonko committed:
refactor: add migration runner interface
